File: implementations/httpcat-python/httpcat/cmd.py
import asyncio
from typing import Annotated

import typer

from .logic import (
    PetType,
    fetch_pet_code_image_as_bytes,
    convert_image_data_to_ascii,
    is_status_code_valid,
)


async def async_cmd_main(*, status_code: int, dog: bool, width: int):
    pet_type = PetType.Cat
    if dog:
        pet_type = PetType.Dog

    data = await fetch_pet_code_image_as_bytes(pet_type, status_code)
    print(convert_image_data_to_ascii(data, width))


def cmd_main(
    status_code: Annotated[int, typer.Argument(help="HTTP status code")],
    dog: Annotated[bool, typer.Option(help="use if you prefer dogs")] = False,
    width: Annotated[int, typer.Option(help="image width")] = 100,
):
    if not is_status_code_valid(status_code):
        print(f"Invalid status code: {status_code}")
        raise typer.Exit(1)

    asyncio.run(async_cmd_main(status_code=status_code, dog=dog, width=width))

File: implementations/httpcat-node/src/cmd.ts
import yargs from "yargs";
import { hideBin } from "yargs/helpers";

import { fetchPetCodeImage, PetType } from "./logic";
import { convertImageToAscii } from "./logic";
import { isStatusCodeValid } from "./logic";

export async function runCmd() {
	const args = await yargs(hideBin(process.argv))
		.scriptName("httpcat-node")
		.usage("Usage: $0 status_code [--dog] [--width image_width]")
		.positional("status_code", {
			describe: "HTTP status code",
		})
		.option("dog", {
			type: "boolean",
			describe: "Use if you prefer dogs",
		})
		.option("width", {
			type: "number",
			describe: "Image width",
			default: 75,
		})
		.demandCommand(1)
		.parse();

	let petType = PetType.Cat;
	if (args.dog) {
		petType = PetType.Dog;
	}

	const statusCode = Number.parseInt(args._[0] as string);
	if (!isStatusCodeValid(statusCode)) {
		console.error(`Invalid status code: ${statusCode}`);
		process.exit(1);
	}

	const imageData = await fetchPetCodeImage(petType, statusCode);
	console.log(await convertImageToAscii(imageData, args.width));
}
